Kernel/Syscalls/iwatcher.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/API/InodeWatcherFlags.h>
+#include <Kernel/FileSystem/Custody.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <Kernel/FileSystem/InodeWatcher.h>
+#include <Kernel/Process.h>
+
+namespace Kernel {
+
+KResultOr<int> Process::sys$create_iwatcher(u32 flags)
+{
+    REQUIRE_PROMISE(rpath);
+
+    int fd = alloc_fd();
+    if (fd < 0)
+        return fd;
+
+    auto description = FileDescription::create(*InodeWatcher::create());
+    if (description.is_error())
+        return description.error();
+
+    m_fds[fd].set(description.release_value());
+    m_fds[fd].description()->set_readable(true);
+
+    // NOTE: these need to happen after the fd is put in the process fd table.
+    if (flags != 0) {
+        auto flags_or_error = sys$fcntl(fd, F_GETFL, 0);
+        if (flags_or_error.is_error())
+            return flags_or_error.error();
+
+        int fd_flags = flags_or_error.value();
+        if (flags & static_cast<unsigned>(InodeWatcherFlags::Nonblock))
+            fd_flags |= O_NONBLOCK;
+        if (flags & static_cast<unsigned>(InodeWatcherFlags::CloseOnExec))
+            fd_flags |= O_CLOEXEC;
+
+        auto error = sys$fcntl(fd, F_SETFL, fd_flags);
+        if (error.is_error())
+            return error.error();
+    }
+
+    return fd;
+}
+
+KResultOr<int> Process::sys$iwatcher_add_watch(Userspace<const Syscall::SC_iwatcher_add_watch_params*> user_params)
+{
+    REQUIRE_PROMISE(rpath);
+
+    Syscall::SC_iwatcher_add_watch_params params;
+    if (!copy_from_user(&params, user_params))
+        return EFAULT;
+
+    auto description = file_description(params.fd);
+    if (!description)
+        return EBADF;
+    if (!description->is_iwatcher())
+        return EBADF;
+    auto iwatcher = description->iwatcher();
+
+    auto path = get_syscall_path_argument(params.user_path.characters, params.user_path.length);
+    if (path.is_error())
+        return path.error();
+
+    auto custody_or_error = VFS::the().resolve_path(path.value(), current_directory());
+    if (custody_or_error.is_error())
+        return custody_or_error.error();
+
+    auto& custody = custody_or_error.value();
+    if (!custody->inode().fs().supports_watchers())
+        return ENOTSUP;
+
+    int wd = alloc_fd();
+    if (wd < 0)
+        return wd;
```

Comment:
I don't think watch descriptors need to be file descriptors.

Instead, they can just be any integer that's unique within the `InodeWatcher`.
Making watch descriptors into full `File` objects seems to unnecessarily complicate a lot of this code.

The way that events are doubly queued (both in `InodeWatcher` and in `WatchDescriptor`) is quite confusing as well.

I'm not sure I understand why the per-descriptor event queues are useful. What use-case do they enable that was not possible without them?

Reply:
Yes, I think the use-case is too niche at the moment. Let's not build that feature until we actually need it. :)